test/core.jl

```diff
@@ -111,6 +111,18 @@ let abcd = ABCDconst(1, 2, 3, 4)
         abcd.d = nothing)
     @test (1, 2, "not constant", 4) === (abcd.a, abcd.b, abcd.c, abcd.d)
 end
+@test begin
+    # Issue #52686
+    for i in 1:2
+        @eval begin
+            struct A{T} end
+            struct B{T, S}
+              a::A{<:T}
+            end
+        end
+    end
+    true
+end
```

Comment:
```suggestion
# Issue #52686
struct A52686{T} end
struct B52686{T, S}
    a::A52686{<:T}
end
function func52686()
    @eval begin
        struct A52686{T} end
        struct B52686{T, S}
            a::A52686{<:T}
        end
    end
    return true
end
@test func52686()
```
I found that the current test case is actually passing on the master. To reproduce and demonstrate the issue fixed by this PR, we need this version.